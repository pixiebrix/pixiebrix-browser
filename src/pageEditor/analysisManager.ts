/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import BlockTypeAnalysis from "@/analysis/analysisVisitors/blockTypeAnalysis";
import ExtensionUrlPatternAnalysis from "@/analysis/analysisVisitors/extensionUrlPatternAnalysis";
import OutputKeyAnalysis from "@/analysis/analysisVisitors/outputKeyAnalysis";
import RenderersAnalysis from "@/analysis/analysisVisitors/renderersAnalysis";
import TemplateAnalysis from "@/analysis/analysisVisitors/templateAnalysis";
import TraceAnalysis from "@/analysis/analysisVisitors/traceAnalysis";
import ReduxAnalysisManager from "@/analysis/ReduxAnalysisManager";
import { type UUID } from "@/core";
import { type TraceRecord } from "@/telemetry/trace";
import { isAnyOf, type PayloadAction } from "@reduxjs/toolkit";
import { type RootState } from "./pageEditorTypes";
import { actions as editorActions } from "@/pageEditor/slices/editorSlice";
import runtimeSlice from "./slices/runtimeSlice";
import RequestPermissionAnalysis from "@/analysis/analysisVisitors/requestPermissionAnalysis";
import FormBrickAnalysis from "@/analysis/analysisVisitors/formBrickAnalysis";
import { selectActiveElementTraces } from "./slices/runtimeSelectors";
import VarAnalysis from "@/analysis/analysisVisitors/varAnalysis/varAnalysis";
import analysisSlice from "@/analysis/analysisSlice";
import RegexAnalysis from "@/analysis/analysisVisitors/regexAnalysis";

const runtimeActions = runtimeSlice.actions;

const pageEditorAnalysisManager = new ReduxAnalysisManager();

// These actions will be used with every analysis so the annotation path is up-to-date
// with the node position in the pipeline
const nodeListMutationActions = [
  editorActions.addNode,
  editorActions.moveNode,
  editorActions.removeNode,
] as const;

// The order in which the analysis are registered is important.
// The first analysis registered will be the first to run.
// When multiple actions (e.g. typing) trigger analysis, the later analysis have more chances to get aborted.
// Try to put the faster analysis first, and the slower ones at the end.

pageEditorAnalysisManager.registerAnalysisEffect(
  (
    action: PayloadAction<{ extensionId: UUID; records: TraceRecord[] }>,
    state: RootState
  ) => {
    // TraceAnalysis filter the trace errors, thus
    // selecting all records here to avoid double filtering
    const records = selectActiveElementTraces(state);

    return new TraceAnalysis(records);
  },
  {
    matcher: isAnyOf(
      runtimeActions.setExtensionTrace,
      ...nodeListMutationActions
    ),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new BlockTypeAnalysis(),
  {
    matcher: isAnyOf(...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new FormBrickAnalysis(),
  {
    matcher: isAnyOf(...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new RenderersAnalysis(),
  {
    matcher: isAnyOf(...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(() => new TemplateAnalysis(), {
  matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
});

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new ExtensionUrlPatternAnalysis(),
  {
    matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new RequestPermissionAnalysis(),
  {
    matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(() => new RegexAnalysis(), {
  matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
});

const varAnalysisFactory = (
  action: PayloadAction<{ extensionId: UUID; records: TraceRecord[] }>,
  state: RootState
) => {
  const records = selectActiveElementTraces(state);
  return new VarAnalysis(records);
};

// OutputKeyAnalysis seems to be the slowest one, so we register it in the end
pageEditorAnalysisManager.registerAnalysisEffect(
  () => new OutputKeyAnalysis(),
  {
    matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
  }
);

// VarAnalysis is not the slowest itself, but it triggers a post-analysis action,
// so it is the last one
pageEditorAnalysisManager.registerAnalysisEffect(
  varAnalysisFactory,
  {
    matcher: isAnyOf(
      editorActions.editElement,
      runtimeActions.setExtensionTrace,
      ...nodeListMutationActions
    ),
  },
  {
    postAnalysisAction(analysis, extensionId, listenerApi) {
      listenerApi.dispatch(
        analysisSlice.actions.setKnownVars({
          extensionId,
          vars: analysis.getKnownVars(),
        })
      );
    },
  }
);

export default pageEditorAnalysisManager;
