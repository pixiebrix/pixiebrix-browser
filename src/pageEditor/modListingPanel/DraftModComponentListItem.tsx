/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./Entry.module.scss";
import React, { useCallback } from "react";
import { actions } from "@/pageEditor/store/editor/editorSlice";
import { useDispatch, useSelector } from "react-redux";
import { ListGroup } from "react-bootstrap";
import { getLabel } from "@/pageEditor/modListingPanel/common";
import {
  ModComponentIcon,
  NotAvailableIcon,
  UnsavedChangesIcon,
} from "@/pageEditor/modListingPanel/ModComponentIcons";
import { type UUID } from "@/types/stringTypes";
import {
  disableOverlay,
  enableOverlay,
  updateSidebar,
} from "@/contentScript/messenger/api";
import { openSidePanel } from "@/utils/sidePanelUtils";
import cx from "classnames";
import reportEvent from "@/telemetry/reportEvent";
import { Events } from "@/telemetry/events";
import { selectSessionId } from "@/pageEditor/store/session/sessionSelectors";
import { type ModComponentFormState } from "@/pageEditor/starterBricks/formStateTypes";
import {
  selectActiveModComponentFormState,
  selectActiveModId,
  selectModComponentIsDirty,
} from "@/pageEditor/store/editor/editorSelectors";
import ModComponentActionMenu from "@/pageEditor/modListingPanel/ModComponentActionMenu";
import { inspectedTab } from "@/pageEditor/context/connection";
import { StarterBrickTypes } from "@/types/starterBrickTypes";

type DraftModComponentListItemProps = {
  modComponentFormState: ModComponentFormState;
  isAvailable: boolean;
  isNested?: boolean;
};

/**
 * A sidebar menu entry corresponding to a touched mod component
 * @see ActivatedModComponentListItem
 * @see ModComponentListItem
 */
const DraftModComponentListItem: React.FunctionComponent<
  DraftModComponentListItemProps
> = ({ modComponentFormState, isAvailable, isNested = false }) => {
  const dispatch = useDispatch();
  const sessionId = useSelector(selectSessionId);
  const activeModId = useSelector(selectActiveModId);
  const activeModComponentFormState = useSelector(
    selectActiveModComponentFormState,
  );

  const isActive =
    activeModComponentFormState?.uuid === modComponentFormState.uuid;
  const modId = modComponentFormState.modMetadata.id;
  const isSiblingOfActiveListItem = activeModComponentFormState?.modMetadata.id
    ? modId === activeModComponentFormState?.modMetadata.id
    : false;
  const isChildOfActiveListItem = modId === activeModId;
  const isRelativeOfActiveListItem =
    !isActive && (isChildOfActiveListItem || isSiblingOfActiveListItem);
  const isDirty = useSelector(
    selectModComponentIsDirty(modComponentFormState.uuid),
  );
  const isButton =
    modComponentFormState.starterBrick.definition.type ===
    StarterBrickTypes.BUTTON;

  const showOverlay = useCallback(async (uuid: UUID) => {
    await enableOverlay(inspectedTab, `[data-pb-uuid="${uuid}"]`);
  }, []);

  const hideOverlay = useCallback(async () => {
    await disableOverlay(inspectedTab);
  }, []);

  return (
    <ListGroup.Item
      className={cx(styles.root, {
        [styles.modBackground ?? ""]: isRelativeOfActiveListItem,
      })}
      as="div"
      active={isActive}
      key={`draft-${modComponentFormState.uuid}`}
      onMouseEnter={
        isButton
          ? async () => showOverlay(modComponentFormState.uuid)
          : undefined
      }
      onMouseLeave={isButton ? async () => hideOverlay() : undefined}
      onClick={async () => {
        reportEvent(Events.PAGE_EDITOR_OPEN, {
          sessionId,
          modComponentId: modComponentFormState.uuid,
        });

        dispatch(actions.setActiveModComponentId(modComponentFormState.uuid));

        if (
          modComponentFormState.starterBrick.definition.type ===
          StarterBrickTypes.SIDEBAR_PANEL
        ) {
          // Switch the sidepanel over to the panel. However, don't refresh because the user might be switching
          // frequently between mod components within the same mod.
          await openSidePanel(inspectedTab.tabId);
          updateSidebar(inspectedTab, {
            modComponentId: modComponentFormState.uuid,
            force: true,
            refresh: false,
          });
        }
      }}
    >
      <span
        className={cx(styles.icon, {
          [styles.nested ?? ""]: isNested,
        })}
      >
        <ModComponentIcon
          type={modComponentFormState.starterBrick.definition.type}
        />
      </span>
      <span className={styles.name}>{getLabel(modComponentFormState)}</span>
      {!isAvailable && (
        <span className={styles.icon}>
          <NotAvailableIcon />
        </span>
      )}
      {isDirty && (
        <span className={cx(styles.icon, styles.unsaved, "text-danger")}>
          <UnsavedChangesIcon />
        </span>
      )}
      {isActive && (
        <ModComponentActionMenu
          modComponentFormState={modComponentFormState}
          labelRoot={getLabel(modComponentFormState)}
          onMoveToMod={async () => {
            dispatch(actions.showMoveCopyToModModal({ moveOrCopy: "move" }));
          }}
        />
      )}
    </ListGroup.Item>
  );
};

export default DraftModComponentListItem;
