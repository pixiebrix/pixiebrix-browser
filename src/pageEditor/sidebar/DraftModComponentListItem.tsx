/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./Entry.module.scss";
import React, { useCallback } from "react";
import { actions } from "@/pageEditor/slices/editorSlice";
import { useDispatch, useSelector } from "react-redux";
import { ListGroup } from "react-bootstrap";
import { getLabel } from "@/pageEditor/sidebar/common";
import {
  ExtensionIcon,
  NotAvailableIcon,
  UnsavedChangesIcon,
} from "@/pageEditor/sidebar/ExtensionIcons";
import { type UUID } from "@/types/stringTypes";
import {
  disableOverlay,
  enableOverlay,
  updateSidebar,
} from "@/contentScript/messenger/api";
import { openSidePanel } from "@/utils/sidePanelUtils";
import cx from "classnames";
import reportEvent from "@/telemetry/reportEvent";
import { Events } from "@/telemetry/events";
import { selectSessionId } from "@/pageEditor/slices/sessionSelectors";
import { type ModComponentFormState } from "@/pageEditor/starterBricks/formStateTypes";
import {
  selectActiveModComponentFormState,
  selectActiveModId,
  selectModComponentIsDirty,
} from "@/pageEditor/slices/editorSelectors";
import ActionMenu from "@/pageEditor/sidebar/ActionMenu";
import useSaveStandaloneModComponent from "@/pageEditor/hooks/useSaveStandaloneModComponent";
import useResetExtension from "@/pageEditor/hooks/useResetExtension";
import {
  useRemoveModComponentFromStorage,
  DEACTIVATE_MOD_MODAL_PROPS,
  DELETE_STANDALONE_MOD_COMPONENT_MODAL_PROPS,
  DELETE_STARTER_BRICK_MODAL_PROPS,
} from "@/pageEditor/hooks/useRemoveModComponentFromStorage";
import useSaveMod from "@/pageEditor/hooks/useSaveMod";
import { selectIsModComponentSavedOnCloud } from "@/store/extensionsSelectors";
import { inspectedTab } from "@/pageEditor/context/connection";

type DraftModComponentListItemProps = {
  modComponentFormState: ModComponentFormState;
  isAvailable: boolean;
  isNested?: boolean;
};

/**
 * A sidebar menu entry corresponding to a touched mod component
 * @see ActivatedModComponentListItem
 * @see ModComponentListItem
 */
const DraftModComponentListItem: React.FunctionComponent<
  DraftModComponentListItemProps
> = ({ modComponentFormState, isAvailable, isNested = false }) => {
  const dispatch = useDispatch();
  const sessionId = useSelector(selectSessionId);
  const activeModId = useSelector(selectActiveModId);
  const activeModComponentFormState = useSelector(
    selectActiveModComponentFormState,
  );

  const isActive =
    activeModComponentFormState?.uuid === modComponentFormState.uuid;
  const modId = modComponentFormState.recipe?.id;
  const isSiblingOfActiveListItem = activeModComponentFormState?.recipe?.id
    ? modId === activeModComponentFormState?.recipe?.id
    : false;
  const isChildOfActiveListItem = modId === activeModId;
  const isRelativeOfActiveListItem =
    !isActive && (isChildOfActiveListItem || isSiblingOfActiveListItem);
  const isDirty = useSelector(
    selectModComponentIsDirty(modComponentFormState.uuid),
  );
  const isSavedOnCloud = useSelector(
    selectIsModComponentSavedOnCloud(modComponentFormState.uuid),
  );
  const removeModComponentFromStorage = useRemoveModComponentFromStorage();
  const isButton = modComponentFormState.type === "menuItem";

  const showOverlay = useCallback(async (uuid: UUID) => {
    await enableOverlay(inspectedTab, `[data-pb-uuid="${uuid}"]`);
  }, []);

  const hideOverlay = useCallback(async () => {
    await disableOverlay(inspectedTab);
  }, []);

  const {
    save: saveStandaloneModComponent,
    isSaving: isSavingStandaloneModComponent,
  } = useSaveStandaloneModComponent();
  const resetExtension = useResetExtension();
  const { save: saveRecipe, isSaving: isSavingRecipe } = useSaveMod();

  const deleteModComponent = async () =>
    removeModComponentFromStorage({
      modComponentId: modComponentFormState.uuid,
      showConfirmationModal: modId
        ? DELETE_STARTER_BRICK_MODAL_PROPS
        : DELETE_STANDALONE_MOD_COMPONENT_MODAL_PROPS,
    });
  const deactivateModComponent = async () =>
    removeModComponentFromStorage({
      modComponentId: modComponentFormState.uuid,
      showConfirmationModal: DEACTIVATE_MOD_MODAL_PROPS,
    });

  const onSave = async () => {
    if (modComponentFormState.recipe) {
      await saveRecipe(modComponentFormState.recipe?.id);
    } else {
      await saveStandaloneModComponent(modComponentFormState);
    }
  };

  const isSaving = modComponentFormState.recipe
    ? isSavingRecipe
    : isSavingStandaloneModComponent;

  const onReset = async () =>
    resetExtension({ extensionId: modComponentFormState.uuid });

  const onDelete = modId || !isSavedOnCloud ? deleteModComponent : undefined;

  const onDeactivate = onDelete ? undefined : deactivateModComponent;

  const onClone = async () => {
    dispatch(actions.cloneActiveExtension());
  };

  return (
    <ListGroup.Item
      className={cx(styles.root, {
        [styles.recipeBackground ?? ""]: isRelativeOfActiveListItem,
      })}
      as="div"
      active={isActive}
      key={`draft-${modComponentFormState.uuid}`}
      onMouseEnter={
        isButton
          ? async () => showOverlay(modComponentFormState.uuid)
          : undefined
      }
      onMouseLeave={isButton ? async () => hideOverlay() : undefined}
      onClick={async () => {
        reportEvent(Events.PAGE_EDITOR_OPEN, {
          sessionId,
          extensionId: modComponentFormState.uuid,
        });

        dispatch(
          actions.makeModComponentFormStateActive(modComponentFormState.uuid),
        );

        if (modComponentFormState.type === "actionPanel") {
          // Switch the sidepanel over to the panel. However, don't refresh because the user might be switching
          // frequently between extensions within the same blueprint.
          await openSidePanel(inspectedTab.tabId);
          updateSidebar(inspectedTab, {
            extensionId: modComponentFormState.uuid,
            force: true,
            refresh: false,
          });
        }
      }}
    >
      <span
        className={cx(styles.icon, {
          [styles.nested]: isNested,
        })}
      >
        <ExtensionIcon type={modComponentFormState.type} />
      </span>
      <span className={styles.name}>{getLabel(modComponentFormState)}</span>
      {!isAvailable && (
        <span className={styles.icon}>
          <NotAvailableIcon />
        </span>
      )}
      {isDirty && !isActive && (
        <span className={cx(styles.icon, styles.unsaved, "text-danger")}>
          <UnsavedChangesIcon />
        </span>
      )}
      {isActive && (
        <ActionMenu
          labelRoot={`${getLabel(modComponentFormState)}`}
          onSave={onSave}
          onDelete={onDelete}
          onDeactivate={onDeactivate}
          onClone={onClone}
          onReset={modComponentFormState.installed ? onReset : undefined}
          isDirty={isDirty}
          onAddToRecipe={
            modComponentFormState.recipe
              ? undefined
              : async () => {
                  dispatch(actions.showAddToRecipeModal());
                }
          }
          onRemoveFromRecipe={
            modComponentFormState.recipe
              ? async () => {
                  dispatch(actions.showRemoveFromRecipeModal());
                }
              : undefined
          }
          disabled={isSaving}
        />
      )}
    </ListGroup.Item>
  );
};

export default DraftModComponentListItem;
