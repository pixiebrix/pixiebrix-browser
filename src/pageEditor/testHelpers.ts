/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import {
  createRenderHookWithWrappers,
  createRenderWithWrappers,
} from "@/testUtils/testHelpers";
import analysisSlice from "@/analysis/analysisSlice";
import pageEditorAnalysisManager from "./analysisManager";
import { tabStateSlice } from "@/pageEditor/tabState/tabStateSlice";
import { appApi } from "@/services/api";
import { modDefinitionsSlice } from "@/modDefinitions/modDefinitionsSlice";
import { modDefinitionsMiddleware } from "@/modDefinitions/modDefinitionsListenerMiddleware";

const configureStoreForTests = () =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      analysis: analysisSlice.reducer,
      tabState: tabStateSlice.reducer,
      modDefinitions: modDefinitionsSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
    middleware(getDefaultMiddleware) {
      /* eslint-disable unicorn/prefer-spread -- It's not Array#concat, can't use spread */
      return getDefaultMiddleware()
        .concat(appApi.middleware)
        .concat(pageEditorAnalysisManager.middleware)
        .concat(modDefinitionsMiddleware);
      /* eslint-enable unicorn/prefer-spread */
    },
  });

const renderWithWrappers = createRenderWithWrappers(configureStoreForTests);
const renderHookWithWrappers = createRenderHookWithWrappers(
  configureStoreForTests
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
export { renderHookWithWrappers as renderHook };
