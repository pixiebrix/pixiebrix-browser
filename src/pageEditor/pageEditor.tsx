/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/bootstrapWithoutRem.css";
import "@/utils/global.scss";
import "@/utils/layout.scss";

import "@/extensionContext";
import "@/development/darkMode.js";

import { initMessengerLogging } from "@/development/messengerLogging";

import ReactDOM from "react-dom";
import React from "react";
import Panel from "@/pageEditor/Panel";
import { watchNavigation } from "@/pageEditor/protocol";
import { initToaster } from "@/utils/notify";
import { initRuntimeLogging } from "@/development/runtimeLogging";
import { initPerformanceMonitoring } from "@/telemetry/performance";
import { setPlatform } from "@/platform/platformContext";
import extensionPagePlatform from "@/extensionPages/extensionPagePlatform";

async function init() {
  setPlatform(extensionPagePlatform);
  void initMessengerLogging();
  void initRuntimeLogging();
  try {
    await initPerformanceMonitoring();
  } catch (error) {
    console.error("Failed to initialize performance monitoring", error);
  }

  watchNavigation();
  initToaster();

  ReactDOM.render(<Panel />, document.querySelector("#container"));
}

void init();
