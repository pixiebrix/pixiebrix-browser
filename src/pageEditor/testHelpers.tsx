/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import { appApi } from "@/services/api";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
