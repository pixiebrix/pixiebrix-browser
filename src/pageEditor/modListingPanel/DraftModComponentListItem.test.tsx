/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render } from "@/pageEditor/testHelpers";
import DraftModComponentListItem from "@/pageEditor/modListingPanel/DraftModComponentListItem";
import { actions as editorActions } from "@/pageEditor/store/editor/editorSlice";
import { authActions } from "@/auth/authSlice";
import { formStateFactory } from "@/testUtils/factories/pageEditorFactories";
import { authStateFactory } from "@/testUtils/factories/authFactories";

jest.mock("@/modDefinitions/modDefinitionHooks", () => ({
  useAllModDefinitions: jest
    .fn()
    .mockReturnValue({ data: [], isLoading: false }),
}));

beforeAll(() => {
  // When a FontAwesomeIcon gets a title, it generates a random id, which breaks the snapshot.
  jest.spyOn(global.Math, "random").mockImplementation(() => 0);
});

afterAll(() => {
  jest.clearAllMocks();
});

describe("DraftModComponentListItem", () => {
  beforeEach(() => {
    // :barf: these Jest snapshots contain sequence UUIDs
    formStateFactory.resetSequence();
  });

  it("renders not active element", () => {
    const formState = formStateFactory();
    expect(
      render(
        <DraftModComponentListItem
          modComponentFormState={formState}
          isAvailable
        />,
        {
          initialValues: formState,
          setupRedux(dispatch) {
            dispatch(authActions.setAuth(authStateFactory()));
            // The addElement also sets the active element
            dispatch(
              editorActions.addModComponentFormState(formStateFactory()),
            );

            // Add new element to deactivate the previous one
            dispatch(editorActions.addModComponentFormState(formState));
            // Remove the active element and stay with one inactive item
            dispatch(
              editorActions.markModComponentFormStateAsDeleted(formState.uuid),
            );
          },
        },
      ).asFragment(),
    ).toMatchSnapshot();
  });

  it("renders active element", () => {
    const formState = formStateFactory();
    expect(
      render(
        <DraftModComponentListItem
          modComponentFormState={formState}
          isAvailable
        />,
        {
          initialValues: formState,
          setupRedux(dispatch) {
            dispatch(authActions.setAuth(authStateFactory()));
            // The addElement also sets the active element
            dispatch(editorActions.addModComponentFormState(formState));
          },
        },
      ).asFragment(),
    ).toMatchSnapshot();
  });
});
