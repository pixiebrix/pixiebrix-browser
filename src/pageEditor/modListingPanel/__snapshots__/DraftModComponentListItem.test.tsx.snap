// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraftModComponentListItem renders active element 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="root list-group-item active"
    >
      <span
        class="icon"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-mouse-pointer fa-w-10 fa-fw "
          data-icon="mouse-pointer"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M302.189 329.126H196.105l55.831 135.993c3.889 9.428-.555 19.999-9.444 23.999l-49.165 21.427c-9.165 4-19.443-.571-23.332-9.714l-53.053-129.136-86.664 89.138C18.729 472.71 0 463.554 0 447.977V18.299C0 1.899 19.921-6.096 30.277 5.443l284.412 292.542c11.472 11.179 3.007 31.141-12.5 31.141z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        class="name"
      >
        Element 1
      </span>
      <div
        class="root"
      >
        <button
          aria-label="Element 1 - Save"
          class="button save btn btn-primary"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-save fa-w-14 "
            data-icon="save"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M433.941 129.941l-83.882-83.882A48 48 0 0 0 316.118 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V163.882a48 48 0 0 0-14.059-33.941zM224 416c-35.346 0-64-28.654-64-64 0-35.346 28.654-64 64-64s64 28.654 64 64c0 35.346-28.654 64-64 64zm96-304.52V212c0 6.627-5.373 12-12 12H76c-6.627 0-12-5.373-12-12V108c0-6.627 5.373-12 12-12h228.52c3.183 0 6.235 1.264 8.485 3.515l3.48 3.48A11.996 11.996 0 0 1 320 111.48z"
              fill="currentColor"
            />
          </svg>
        </button>
        <span>
          <button
            aria-expanded="false"
            aria-haspopup="true"
            aria-label="Element 1 - Ellipsis"
            class="szh-menu-button button ellipsisMenu"
            data-testid="ellipsis-menu-button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-ellipsis-v fa-w-6 "
              data-icon="ellipsis-v"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 192 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                fill="currentColor"
              />
            </svg>
          </button>
        </span>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`DraftModComponentListItem renders not active element 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="root list-group-item"
    >
      <span
        class="icon"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-mouse-pointer fa-w-10 fa-fw "
          data-icon="mouse-pointer"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M302.189 329.126H196.105l55.831 135.993c3.889 9.428-.555 19.999-9.444 23.999l-49.165 21.427c-9.165 4-19.443-.571-23.332-9.714l-53.053-129.136-86.664 89.138C18.729 472.71 0 463.554 0 447.977V18.299C0 1.899 19.921-6.096 30.277 5.443l284.412 292.542c11.472 11.179 3.007 31.141-12.5 31.141z"
            fill="currentColor"
          />
        </svg>
      </span>
      <span
        class="name"
      >
        Element 1
      </span>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
